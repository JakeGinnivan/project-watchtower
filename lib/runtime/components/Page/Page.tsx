import * as React from 'react'
import { PageLifecycle } from '../PageLifecycleProvider/PageLifecycleProvider'
import { Logger } from 'pino'

export interface State { error: boolean }
export interface Props {
    errorComponent: React.ReactType
}

export class Page extends React.Component<Props, State> {
    static contextTypes = {
        // Seems like context cannot be exported, this is a runtime react thing anyways
        pageLifecycle: React.PropTypes.object as any,
    }

    state: State = { error: false }

    context: {
        pageLifecycle: PageLifecycle,
        logger: Logger,
    }

    // This is using React error boundaries which were added in v15
    // v16 may add official support
    // Info at:
    // tslint:disable-next-line:max-line-length
    // https://github.com/facebook/react/blob/15-stable/src/renderers/shared/stack/reconciler/tests/ReactErrorBoundaries-test.js
    // https://github.com/facebook/react/issues/2461
    // TODO Move to stable error boundaries API
    unstable_handleError(e: Error) {
        e.message = 'Render error: ' + e.message
        this.context.logger.error({ err: e })
        this.setState({ error: true })
    }

    componentDidMount() {
        this.context.pageLifecycle.pageRenderComplete()
        this.context.pageLifecycle.onRouteChanged(this.routeChanged)
    }

    componentWillUnmount() {
        this.context.pageLifecycle.offRouteChanged(this.routeChanged)
    }

    routeChanged = () => {
        if (this.state.error) {
            this.setState({ error: false })
        }
    }

    componentDidUpdate() {
        this.context.pageLifecycle.pageRenderComplete()
    }

    render() {
        if (this.state.error) {
            return (
                <div className="js-page-contents">
                    <this.props.errorComponent />
                </div>
            )
        }

        return (
            <div className="js-page-contents">
                {this.props.children}
            </div>
        )
    }
}
