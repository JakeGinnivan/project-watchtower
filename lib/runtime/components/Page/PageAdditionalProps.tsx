import * as React from 'react'
import * as PropTypes from 'prop-types'
import { PageLifecycleProps } from '../PageLifecycleProvider/PageLifecycleProvider'
import { Logger } from '../../util/log'
import { PageLifecycle } from '../PageLifecycleProvider/PageLifecycle'

export interface Props {
    pageProperties?: object
}

export class PageAdditionalProps extends React.PureComponent<Props & PageLifecycleProps, {}> {
    static contextTypes = {
        // Seems like context cannot be exported, this is a runtime react thing anyways
        pageLifecycle: PropTypes.object as any,
    }

    context: {
        pageLifecycle: PageLifecycle
        logger: Logger
    }

    constructor(
        props: Props & PageLifecycleProps,
        context: {
            pageLifecycle: PageLifecycle
            logger: Logger
        },
    ) {
        super(props, context)

        context.pageLifecycle.updatePageProps(this.props.pageProperties || {})
    }

    componentWillReceiveProps(nextProps: Props & PageLifecycleProps) {
        this.context.pageLifecycle.updatePageProps(nextProps.pageProperties || {})
    }

    render() {
        return this.props.children ? React.Children.only(this.props.children) : null
    }
}
