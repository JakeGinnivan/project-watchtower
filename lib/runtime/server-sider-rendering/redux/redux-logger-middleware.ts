import * as redux from 'redux'

// @TODO revisit and make logger and action types correct
const reduxLoggerMiddleware = (logger: any) =>
    <S>() =>
        (next: redux.Dispatch<S>) =>
            (action: any) => {
                let actionDescription = action.type
                if (action.meta) {
                    actionDescription = `${actionDescription} > ${JSON.stringify(action.meta)}`
                }
                if (action.payload) {
                    const output = typeof action.payload === 'string' ? action.payload : '{...}'
                    // tslint:disable-next-line max-line-length
                    actionDescription = `${actionDescription} > ${action.payload ? `/w payload: ${output}` : ''}`
                }

                // new error payload strategy
                if (action.isError) {
                    const logObj = { err: action, event: 'ActionFailure' }
                    const logMsg = `Failed to perform client action ${actionDescription}`
                    logger.warn(logObj, logMsg)
                } else if (action.error) { // old error payload strategy
                    const logObj = { err: action, event: 'ActionFailure' }
                    const logMsg = `Failed to perform client action ${actionDescription}`

                    logger.warn(logObj, logMsg)
                }

                logger.debug(`dispatching: ${actionDescription}`)

                const result = next(action)

                logger.debug(`finished processing: ${actionDescription} (including react updates)`)

                return result
            }

export default reduxLoggerMiddleware
