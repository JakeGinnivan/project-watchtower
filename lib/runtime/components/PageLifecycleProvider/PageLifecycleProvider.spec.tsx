/**
 * @jest-environment jsdom
 */

import * as React from 'react'
import { createStore } from 'redux'
import { Provider } from 'react-redux'
import { MemoryRouter, Route } from 'react-router-dom'
import { mount } from 'enzyme'
import * as H from 'history'

import {
    PageLifecycleProvider, withPageLifecycleEvents, ComponentWithLifecycle, PageEvent,
} from './PageLifecycleProvider'
import PromiseCompletionSource from '../../util/promise-completion-source'
import { DataLoaderResources } from 'redux-data-loader'
import { Page } from '../Page/Page'

interface TestData {
    bar: string
}

const loaderStatusLookup: { [state: number]: string } = {
    0: 'Idle',
    1: 'Fetching',
    2: 'Refreshing',
    3: 'Paging',
    4: 'Updating',
}

const createTestComponents = () => {
    const resources = new DataLoaderResources()
    const promiseCompletionSource = new PromiseCompletionSource<TestData>()
    const TestDataLoader = resources.registerResource(
        'test',
        (_id: string) => {
            return promiseCompletionSource.promise
        },
    )
    class TestPage extends React.Component<{ path: string }, {}> {
        render() {
            return (
                <Page errorComponent="oops">
                    <TestDataLoader
                        resourceId={'test' + this.props.path}
                        renderData={(data) => (
                            <div>
                                Status: {loaderStatusLookup[data.status]}
                                Last action: {JSON.stringify(data.lastAction || {})}
                            </div>
                        )}
                    />
                </Page>
            )
        }
    }

    const FakeLazyLoad = withPageLifecycleEvents(
        // tslint:disable-next-line:max-classes-per-file
        class extends ComponentWithLifecycle<{ path: string}, { loaded: boolean }> {
            state = { loaded: false }

            componentDidMount() {
                this.context.pageLifecycle.lazyLoadingPage()
                // When we mount, pretend we are starting to load the test page
                setTimeout(() => {
                    this.context.pageLifecycle.lazyLoadingPageComplete()
                    this.setState({ loaded: true })
                })
            }

            render() {
                return (
                    this.state.loaded
                        ? <TestPage {...this.props} />
                        : <noscript />
                )
            }
        },
    )
    return {
        resources,
        promiseCompletionSource,
        TestDataLoader,
        FakeLazyLoad,
        TestPage,
    }
}

describe('PageLifecycleProvider', () => {
    it('raises loading event after render', () => {
        const testComponents = createTestComponents()
        const pageEvents: PageEvent[] = []

        const store = createStore((s = {}, _) => s)
        const wrapper = mount((
            <Provider store={store}>
                <MemoryRouter initialEntries={['/']} initialIndex={0}>
                    <PageLifecycleProvider
                        resources={testComponents.resources}
                        onEvent={(event) => pageEvents.push(event)}
                    >
                        <testComponents.TestPage path="/" />
                    </PageLifecycleProvider>
                </MemoryRouter>
            </Provider>
        ))

        const testPage = wrapper.find(testComponents.TestPage)

        expect(pageEvents.map((e) => {
            e.timeStamp = 0
            return e
        })).toMatchSnapshot()
        expect(testPage.debug()).toMatchSnapshot()
    })

    it('raises completed event after data loaded and content re-rendered', async () => {
        const testComponents = createTestComponents()
        const pageEvents: PageEvent[] = []

        const store = createStore((s = {}, _) => s)
        const wrapper = mount((
            <Provider store={store}>
                <MemoryRouter initialEntries={['/']} initialIndex={0}>
                    <PageLifecycleProvider
                        resources={testComponents.resources}
                        onEvent={(event) => pageEvents.push(event)}
                    >
                        <testComponents.TestPage path="/" />
                    </PageLifecycleProvider>
                </MemoryRouter>
            </Provider>
        ))

        const testPage = wrapper.find(testComponents.TestPage)
        testComponents.promiseCompletionSource.resolve({
            bar: 'test',
        })
        expect(pageEvents.length).toBe(1)
        await new Promise((resolve) => setTimeout(() => resolve()))

        expect(pageEvents.map((e) => {
            e.timeStamp = 0
            return e
        })).toMatchSnapshot()
        expect(testPage.debug()).toMatchSnapshot()
    })

    it('raises completed event after lazy loaded page has completed load data', async () => {
        const testComponents = createTestComponents()
        const pageEvents: PageEvent[] = []

        const store = createStore((s = {}, _) => s)
        const wrapper = mount((
            <Provider store={store}>
                <MemoryRouter initialEntries={['/']} initialIndex={0}>
                    <PageLifecycleProvider
                        resources={testComponents.resources}
                        onEvent={(event) => pageEvents.push(event)}
                    >
                        <Page errorComponent="oops">
                            <testComponents.FakeLazyLoad path="/" />
                        </Page>
                    </PageLifecycleProvider>
                </MemoryRouter>
            </Provider>
        ))

        const testPage = wrapper.find(testComponents.TestPage)

        expect(pageEvents.map((e) => {
            e.timeStamp = 0
            return e
        })).toMatchSnapshot()
        expect(testPage.debug()).toMatchSnapshot()
        await new Promise((resolve) => setTimeout(() => resolve()))

        testComponents.promiseCompletionSource.resolve({
            bar: 'test',
        })
        // We should not have received the completed event yet
        expect(pageEvents.length).toBe(1)

        // Let the completion propagate
        await new Promise((resolve) => setTimeout(() => resolve()))
        expect(pageEvents.map((e) => {
            e.timeStamp = 0
            return e
        })).toMatchSnapshot()
        expect(testPage.debug()).toMatchSnapshot()
    })

    // TODO Test which validates page load events on navigation..

    it('raises events when page navigated', async () => {
        const testComponents = createTestComponents()
        let history: H.History | undefined
        const pageEvents: PageEvent[] = []

        const store = createStore((s = {}, _) => s)
        const wrapper = mount((
            <Provider store={store}>
                <MemoryRouter initialEntries={['/', '/foo']} initialIndex={0}>
                    <PageLifecycleProvider
                        resources={testComponents.resources}
                        onEvent={(event) => pageEvents.push(event)}
                    >
                        <Route
                            render={(props) => {
                                history = props.history
                                return (
                                    <testComponents.TestPage path={props.location.pathname} />
                                )}
                            }
                        />
                    </PageLifecycleProvider>
                </MemoryRouter>
            </Provider>
        ))

        testComponents.promiseCompletionSource.resolve({ bar: 'test' })
        await new Promise((resolve) => setTimeout(() => resolve()))

        if (history) { history.push('/foo') }

        const testPage = wrapper.find(testComponents.TestPage)

        testComponents.promiseCompletionSource.resolve({ bar: 'page2' })
        await new Promise((resolve) => setTimeout(() => resolve()))
        expect(pageEvents.map((e) => {
            e.timeStamp = 0
            return e
        })).toMatchSnapshot()
        expect(testPage.debug()).toMatchSnapshot()
    })
})
