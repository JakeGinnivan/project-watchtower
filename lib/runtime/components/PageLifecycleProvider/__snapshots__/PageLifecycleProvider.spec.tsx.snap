// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageLifecycleProvider can pass extra data to pages 1`] = `
Array [
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-started",
  },
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "home": "isHome",
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "pageExtra": "Some extra page data",
    },
    "timeStamp": 0,
    "type": "page-load-complete",
  },
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "foo": "isFoo",
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/foo",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-started",
  },
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "foo": "isFoo",
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/foo",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-complete",
  },
]
`;

exports[`PageLifecycleProvider can pass extra data to pages 2`] = `
"<TestPage path=\\"/foo\\" extraProps={{...}}>
  <withLog(withPageLifecycleProps(Page)) pageProperties={{...}} page={[Function]}>
    <withPageLifecycleProps(Page) pageProperties={{...}} page={[Function]} logger={{...}}>
      <Page pageProperties={{...}} page={[Function]} logger={{...}} currentPageState=\\"loaded\\" currentPageLocation={{...}} beginLoadingData={[Function]} endLoadingData={[Function]}>
        <div>
          Page location: 
          /foo
          Page state: 
          loaded
        </div>
      </Page>
    </withPageLifecycleProps(Page)>
  </withLog(withPageLifecycleProps(Page))>
</TestPage>"
`;

exports[`PageLifecycleProvider raises completed event after data loaded and content re-rendered 1`] = `
Array [
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-started",
  },
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-complete",
  },
]
`;

exports[`PageLifecycleProvider raises completed event after data loaded and content re-rendered 2`] = `
"<TestPage path=\\"/\\">
  <withLog(withPageLifecycleProps(Page)) pageProperties={[undefined]} page={[Function]}>
    <withPageLifecycleProps(Page) pageProperties={[undefined]} page={[Function]} logger={{...}}>
      <Page pageProperties={[undefined]} page={[Function]} logger={{...}} currentPageState=\\"loaded\\" currentPageLocation={{...}} beginLoadingData={[Function]} endLoadingData={[Function]}>
        <div>
          Page location: 
          /
          Page state: 
          loaded
        </div>
      </Page>
    </withPageLifecycleProps(Page)>
  </withLog(withPageLifecycleProps(Page))>
</TestPage>"
`;

exports[`PageLifecycleProvider raises completed event after lazy loaded page has completed load data 1`] = `
Array [
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-started",
  },
]
`;

exports[`PageLifecycleProvider raises completed event after lazy loaded page has completed load data 2`] = `""`;

exports[`PageLifecycleProvider raises completed event after lazy loaded page has completed load data 3`] = `
Array [
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-started",
  },
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-complete",
  },
]
`;

exports[`PageLifecycleProvider raises completed event after lazy loaded page has completed load data 4`] = `""`;

exports[`PageLifecycleProvider raises events when page navigated 1`] = `
Array [
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-started",
  },
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-complete",
  },
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/foo",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-started",
  },
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/foo",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-complete",
  },
]
`;

exports[`PageLifecycleProvider raises events when page navigated 2`] = `
"<TestPage path=\\"/foo\\">
  <withLog(withPageLifecycleProps(Page)) pageProperties={[undefined]} page={[Function]}>
    <withPageLifecycleProps(Page) pageProperties={[undefined]} page={[Function]} logger={{...}}>
      <Page pageProperties={[undefined]} page={[Function]} logger={{...}} currentPageState=\\"loaded\\" currentPageLocation={{...}} beginLoadingData={[Function]} endLoadingData={[Function]}>
        <div>
          Page location: 
          /foo
          Page state: 
          loaded
        </div>
      </Page>
    </withPageLifecycleProps(Page)>
  </withLog(withPageLifecycleProps(Page))>
</TestPage>"
`;

exports[`PageLifecycleProvider raises loading event after render 1`] = `
Array [
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-started",
  },
]
`;

exports[`PageLifecycleProvider raises loading event after render 2`] = `
"<TestPage path=\\"/\\">
  <withLog(withPageLifecycleProps(Page)) pageProperties={[undefined]} page={[Function]}>
    <withPageLifecycleProps(Page) pageProperties={[undefined]} page={[Function]} logger={{...}}>
      <Page pageProperties={[undefined]} page={[Function]} logger={{...}} currentPageState=\\"loading\\" currentPageLocation={{...}} beginLoadingData={[Function]} endLoadingData={[Function]}>
        <div>
          Page location: 
          /
          Page state: 
          loading
        </div>
      </Page>
    </withPageLifecycleProps(Page)>
  </withLog(withPageLifecycleProps(Page))>
</TestPage>"
`;
